nussl/separation/nmf_mfcc.py

```diff
@@ -0,0 +1,215 @@
+        if isinstance(mfcc_range, int) and mfcc_range < n_mfcc:
+            self.mfcc_start, self.mfcc_end = 1, mfcc_range
+        elif isinstance(mfcc_range, (tuple, list)) and len(mfcc_range) == 2:
+            self.mfcc_start, self.mfcc_end = mfcc_range[0], mfcc_range[1]
+        else:
+            raise ValueError('mfcc_range is not set correctly! Must be a tuple or list with min and max, or int (max)')
+
+        # If kmeans_kwargs does not include the 'random_state', use the random_seed instead. Else, use the value
+        # provided in the dictionary. If kmeans_kwargs is None, use the random_seed.
+        self.kmeans_random_seed = kmeans_kwargs.pop('random_state', random_seed) \
+            if isinstance(kmeans_kwargs, dict) else random_seed
+
+        # Initialize the K Means clusterer
+        if isinstance(self.kmeans_kwargs, dict):
+            self.clusterer = sklearn.cluster.KMeans(n_clusters=self.num_sources, random_state=self.kmeans_random_seed,
+                                                    **kmeans_kwargs)
+        else:
+            self.clusterer = sklearn.cluster.KMeans(n_clusters=self.num_sources, random_state=self.kmeans_random_seed)
+
+    def run(self):
+        """ This function calls TransformerNMF on the magnitude spectrogram of each channel in the input audio signal.
+        The templates and activation matrices returned are clustered using K-Means clustering. These clusters are used
+        to create mask objects for each source. Note: The masks are not returned in a particular order, but they are in
+        the same order for each channel.
+
+        Returns:
+            self.masks (np.array): A list of binary mask objects that can be used to extract the sources.
+
+        Example:
+
+        .. code-block:: python
+            :linenos:
+
+            signal = nussl.AudioSignal(path_to_input_file='input_name.wav')
+
+            # Set up and run NMF MFCC
+            nmf_mfcc =  nussl.NMF_MFCC(signal, num_sources=2) # Returns a binary mask by default
+            masks = nmf_mfcc.run()
+
+            # Get audio signals
+            sources = nmf_mfcc.make_audio_signals()
+
+            # Output the sources
+            for i, source in enumerate(sources):
+                output_file_name = str(i) + '.wav'
+                source.write_audio_to_file(output_file_name)
+        """
+        self.masks = []
+        self.input_audio_signal.stft_params = self.stft_params
+        self.signal_stft = self.input_audio_signal.stft()
+
+        for i in range(self.input_audio_signal.num_channels):
+            channel_stft = self.input_audio_signal.get_magnitude_spectrogram_channel(i)
+
+            # Set up NMF and run
+            nmf = transformer_nmf.TransformerNMF(input_matrix=channel_stft, num_components=self.num_templates,
+                                                 seed=self.random_seed)
+
+            # Pass these to constructor (above)
+            nmf.should_use_epsilon = False
```

Comment:
I was going to say that these variables could all go in the constructor, but it looks like you've already left yourself a comment here to do just that.